src/EditorFeatures/CSharp/EventHookup/EventHookupCommandHandler_TabKeyCommand.cs

```diff
@@ -85,116 +78,179 @@ private void HandleTabWorker(ITextView textView, ITextBuffer subjectBuffer, Acti
             }
         }
 
-        // Blocking wait (if necessary) to determine whether to consume the tab and
-        // generate the event handler.
-        EventHookupSessionManager.CurrentSession.GetEventNameTask.Wait(cancellationToken);
+        var subjectBuffer = args.SubjectBuffer;
+        var caretPoint = args.TextView.GetCaretPoint(subjectBuffer);
+        if (caretPoint is null)
+            return false;
+
+        var currentSnapshot = subjectBuffer.CurrentSnapshot;
+        var document = currentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
+        if (document is null)
+            return false;
+
+        // Now emit the event asynchronously.
+        var token = _asyncListener.BeginAsyncOperation(nameof(ExecuteCommand));
+
+        // Capture everything we need off of the session manager as we'll be dismissing the core session immediately
+        // before we kick off the work to emit hte event.
+        var eventNameTask = EventHookupSessionManager.CurrentSession.GetEventNameTask;
+        var applicableToSpan = EventHookupSessionManager.CurrentSession.TrackingSpan.GetSpan(currentSnapshot);
+
+        // Ensure no matter what that once tab is hit that we're back to the initial no-session state. We do not
+        // want to cancel the bg tasks kicked off as we need their values to actually emit the event.
+        EventHookupSessionManager.DismissExistingSessions(cancelBackgroundTasks: false);
+
+        var task = ExecuteCommandAsync(
+            args,
+            nextHandler,
+            applicableToSpan,
+            document,
+            eventNameTask,
+            caretPoint.Value);
+        task.CompletesAsyncOperation(token);
+
+        // At this point, we've taken control, so don't send the tab into the buffer.  But do dismiss the overall
+        // session.  We no longer need it.
+        return true;
+    }
 
-        string eventHandlerMethodName = null;
-        if (EventHookupSessionManager.CurrentSession.GetEventNameTask.Status == TaskStatus.RanToCompletion)
+    private async Task ExecuteCommandAsync(
+        TabKeyCommandArgs args,
+        Action nextHandler,
+        SnapshotSpan applicableToSpan,
+        Document document,
+        Task<string> getEventNameTask,
+        SnapshotPoint initialCaretPoint)
+    {
+        try
         {
-            eventHandlerMethodName = EventHookupSessionManager.CurrentSession.GetEventNameTask.WaitAndGetResult(cancellationToken);
+            _threadingContext.ThrowIfNotOnUIThread();
+            await ExecuteCommandWorkerAsync().ConfigureAwait(false);
         }
-
-        if (eventHandlerMethodName == null ||
-            EventHookupSessionManager.CurrentSession.TextView != textView)
+        catch (OperationCanceledException)
+        {
+        }
```

Comment:
Should the body of this be calling `DismissExistingSessions(cancelBackgroundTasks: true)`? There are a bunch of different possible resolutions to this.

Reply:
no.  the caller of this method *always* dismisses the initial session first.